File: websocket-gateway/src/main/java/com/realfds/gateway/config/WebSocketConfig.java
package com.realfds.gateway.config;

import com.realfds.gateway.handler.AlertWebSocketHandler;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.socket.config.annotation.EnableWebSocket;
import org.springframework.web.socket.config.annotation.WebSocketConfigurer;
import org.springframework.web.socket.config.annotation.WebSocketHandlerRegistry;

/**
 * WebSocket 설정 클래스
 *
 * WebSocket 엔드포인트를 설정하고 핸들러를 매핑합니다.
 *
 * 주요 설정:
 * - WebSocket 엔드포인트: /ws/alerts
 * - CORS 설정: 모든 Origin 허용 (로컬 개발용)
 * - SockJS fallback 활성화 (브라우저 호환성)
 */
@Configuration
@EnableWebSocket
public class WebSocketConfig implements WebSocketConfigurer {

    private final AlertWebSocketHandler alertWebSocketHandler;

    public WebSocketConfig(AlertWebSocketHandler alertWebSocketHandler) {
        this.alertWebSocketHandler = alertWebSocketHandler;
    }

    @Override
    public void registerWebSocketHandlers(WebSocketHandlerRegistry registry) {
        // WebSocket 핸들러 매핑
        // /ws/alerts 엔드포인트로 WebSocket 연결 가능
        // 모든 Origin 허용 (로컬 개발용 - 프로덕션에서는 제한 필요)
        registry.addHandler(alertWebSocketHandler, "/ws/alerts")
                .setAllowedOrigins("*");  // 로컬 개발용: 모든 Origin 허용
    }
}

File: alert-dashboard/backend/src/main/kotlin/io/realfds/alert/config/WebSocketConfig.kt
package io.realfds.alert.config

import org.springframework.context.annotation.Configuration
import org.springframework.web.socket.config.annotation.EnableWebSocket
import org.springframework.web.socket.config.annotation.WebSocketConfigurer
import org.springframework.web.socket.config.annotation.WebSocketHandlerRegistry
import org.springframework.web.socket.server.support.HttpSessionHandshakeInterceptor

/**
 * WebSocket 설정
 *
 * 실시간 메트릭 데이터 브로드캐스트를 위한 WebSocket 엔드포인트를 등록합니다.
 *
 * ## 엔드포인트
 * - **ws://localhost:8082/ws/metrics**: 실시간 메트릭 데이터 스트림
 *
 * ## 설정 특징
 * - CORS 허용 (로컬 개발용): http://localhost:3000
 * - SockJS fallback 미사용 (순수 WebSocket만 사용)
 * - HTTP 세션 연동 (HttpSessionHandshakeInterceptor)
 *
 * @see io.realfds.alert.websocket.MetricsWebSocketHandler
 */
@Configuration
@EnableWebSocket
class WebSocketConfig(
    private val metricsWebSocketHandler: io.realfds.alert.websocket.MetricsWebSocketHandler
) : WebSocketConfigurer {

    /**
     * WebSocket 핸들러 등록
     *
     * @param registry WebSocket 핸들러 레지스트리
     */
    override fun registerWebSocketHandlers(registry: WebSocketHandlerRegistry) {
        registry
            // 실시간 메트릭 데이터 WebSocket 엔드포인트 등록
            .addHandler(metricsWebSocketHandler, "/ws/metrics")
            // CORS 설정 (로컬 개발용)
            .setAllowedOrigins(
                "http://localhost:3000",      // 로컬 React 개발 서버
                "http://localhost:8080",      // 로컬 프론트엔드 프로덕션
                "http://localhost:8082",      // 백엔드 서버 (동일 origin)
                "http://frontend-dashboard:8080"  // Docker 컨테이너
            )
            // HTTP 세션 정보를 WebSocket 세션에 복사
            .addInterceptors(HttpSessionHandshakeInterceptor())
            // SockJS fallback 미사용 (순수 WebSocket만 사용)
            // .withSockJS()를 호출하지 않음
    }
}
